pcsx2/gui/Dialogs/GSDumpDialog.cpp

```diff
@@ -0,0 +1,860 @@
+			tex_infos[9].Printf("CSM = %s", TEXCSMNames[(u32)((data.lo >> 55) & ((u64)(1 << 1) - 1))]);
+			tex_infos[10].Printf("CSA = %u", (u32)((data.lo >> 56) & ((u64)(1 << 5) - 1)));
+			tex_infos[11].Printf("CLD = %u", (u32)((data.lo >> 61) & ((u64)(1 << 3) - 1)));
+
+			for (auto& el : tex_infos)
+				m_gif_packet->AppendItem(rootId, el);
+			break;
+		}
+		case FOG:
+		{
+			wxString s;
+			if (packed)
+				s.Printf("F = %u", (u32)((data.hi >> 36) & ((u64)(1 << 8) - 1)));
+			else
+				s.Printf("F = %u", (u32)((data.lo >> 56) & ((u64)(1 << 8) - 1)));
+			m_gif_packet->AppendItem(rootId, s);
+			break;
+		}
+		case AD:
+		{
+			GIFReg nreg = (GIFReg)(data.hi & ((u64)(1 << 8) - 1));
+			if ((GIFReg)nreg == AD)
+			{
+				wxString s;
+				s.Printf("NOP");
+				m_gif_packet->AppendItem(id, s);
+			}
+			else
+				ParseTreeReg(id, nreg, data, packed);
+			m_gif_packet->Delete(rootId);
+			break;
+		}
+		default:
+			break;
+	}
+}
+
+void Dialogs::GSDumpDialog::ParseTreePrim(wxTreeItemId& id, u32 prim)
+{
+	std::vector<wxString> prim_infos(9);
+
+	prim_infos[0].Printf("Primitive Type = %s", GsPrimNames[(prim & ((u64)(1 << 3) - 1))]);
+	prim_infos[1].Printf("IIP = %s", GsIIPNames[((prim >> 3) & 1)]);
+	prim_infos[2].Printf("TME = %s", (bool)((prim >> 4) & 1) ? "True" : "False");
+	prim_infos[3].Printf("FGE = %s", (bool)((prim >> 5) & 1) ? "True" : "False");
+	prim_infos[4].Printf("FGE = %s", (bool)((prim >> 6) & 1) ? "True" : "False");
+	prim_infos[5].Printf("AA1 = %s", (bool)((prim >> 7) & 1) ? "True" : "False");
+	prim_infos[6].Printf("FST = %s", GsFSTNames[((prim >> 3) & 1)]);
+	prim_infos[7].Printf("CTXT = %s", GsCTXTNames[((prim >> 9) & 1)]);
+	prim_infos[8].Printf("FIX = %s", GsFIXNames[((prim >> 10) & 1)]);
+
+	for (auto& el : prim_infos)
+		m_gif_packet->AppendItem(id, el);
+}
+
+void Dialogs::GSDumpDialog::CheckDebug(wxCommandEvent& event)
+{
+	if (m_debug_mode->GetValue())
+	{
+		GenPacketList();
+		m_start->Enable();
+		m_step->Enable();
+		m_selection->Enable();
+		m_vsync->Enable();
+	}
+	else
+	{
+		m_gif_list->DeleteAllItems();
+		m_gif_packet->DeleteAllItems();
+		m_gif_list->Refresh();
+		m_start->Disable();
+		m_step->Disable();
+		m_selection->Disable();
+		m_vsync->Disable();
+	}
+}
+
+Dialogs::GSDumpDialog::GSThread::GSThread(GSDumpDialog* dlg)
+	: pxThread("GSDump")
+	, m_root_window(dlg)
+{
+}
+
+Dialogs::GSDumpDialog::GSThread::~GSThread()
+{
+	try
+	{
+		pxThread::Cancel();
+	}
+	DESTRUCTOR_CATCHALL
+}
+
+void Dialogs::GSDumpDialog::GSThread::OnStop()
+{
+	m_root_window->m_debug_mode->Disable();
+	m_root_window->m_start->Disable();
+	m_root_window->m_step->Disable();
+	m_root_window->m_selection->Disable();
+	m_root_window->m_vsync->Disable();
+	m_root_window->m_gif_list->DeleteAllItems();
+	m_root_window->m_gif_packet->DeleteAllItems();
+	m_root_window->m_button_events.clear();
+	m_root_window->m_debug_mode->SetValue(false);
+	m_root_window->m_run->Enable();
+	m_dump_file->Close();
+}
```

Comment:
This function modifies the GUI state from the task thread. Is that safe?

Reply:
As far as I know it is, gif list and packets are generated inside the thread and only inside the thread, all other buttons cannot have their state changed while the thread is running, m_run enables you to start a new thread and is enabled last.